presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java

```diff
@@ -1400,6 +1462,195 @@ public synchronized void progressChanged(ProgressEvent progressEvent)
         }
     }
 
+    private static class PrestoS3StreamingOutputStream
+            extends OutputStream
+    {
+        private final AmazonS3 s3;
+        private final String bucketName;
+        private final String key;
+        private final String uploadId;
+
+        private final ListeningExecutorService uploadExecutor;
+
+        private int currentPartNumber;
+        private byte[] buf;
+        private int count;
+
+        private ListenableFuture<UploadPartResult> inProgressUploadFuture;
+        private final List<UploadPartResult> parts;
+
+        public PrestoS3StreamingOutputStream(
+                AmazonS3 s3,
+                String bucketName,
+                String key,
+                String uploadId,
+                ListeningExecutorService uploadExecutor,
+                long multiPartUploadMinPartSize)
+        {
+            STATS.uploadStarted();
+
+            this.s3 = requireNonNull(s3, "s3 is null");
+
+            int bufferSize = (int) multiPartUploadMinPartSize;
+            this.buf = new byte[bufferSize];
+
+            this.bucketName = bucketName;
+            this.key = key;
+            this.uploadId = uploadId;
+
+            this.uploadExecutor = uploadExecutor;
+
+            this.parts = new ArrayList<>();
+        }
+
+        @Override
+        public void write(int b)
+                throws IOException
+        {
+            if (count >= buf.length) {
+                flushBuffer(false);
+            }
+
+            buf[count] = (byte) b;
+            count++;
+        }
+
+        @Override
+        public void write(byte[] bytes, int off, int len)
+                throws IOException
+        {
+            int offset = off;
+            int remain = len;
+            while ((buf.length - count) <= remain) {
+                int copied = buf.length - count;
+                arraycopy(bytes, offset, buf, count, copied);
+                count += copied;
+
+                flushBuffer(false);
+
+                offset += copied;
+                remain -= copied;
+            }
+
+            if (remain > 0) {
+                arraycopy(bytes, offset, buf, count, remain);
+                count += remain;
+            }
+        }
+
+        @Override
+        public void flush()
+                throws IOException
+        {
+            flushBuffer(false);
+        }
+
+        @Override
+        public void close()
+                throws IOException
+        {
+            try {
+                flushBuffer(true);
+                waitForPreviousUploadFinish();
+            }
+            catch (IOException e) {
+                abortUpload(true);
+                throw e;
+            }
+
+            if (!parts.isEmpty()) {
+                try {
+                    finishUpload(parts);
+                }
+                catch (Exception e) {
```

Comment:
This can catch `RuntimeException` since `finishUpload()` doesn't throw checked exceptions